rfd/0194-k8s-rbac-simplification.md

```diff
@@ -0,0 +1,220 @@
+resources. Granting all of a namespace can be too permissive and granting
+individually `ingress`, `deployments`, `services`, `pods`, etc, etc can be
+cumbersome and error prone.
+
+### UX
+
+#### Deprecate `namespace`
+
+We currently have the `namespace` kind mean "everything within a namespace",
+which has caused confusion with customers and resulting in unexpected access
+granted.
+
+This should be deprecated and removed. It would require a bump on the role
+model to v8 to avoid changing behavior for current customers.
+
+Upgrading to v8 would map the deprecated `namespace` field to `*` while
+swapping `name`/`namespace` as these are equivalent:
+
+```yaml
+kind: namespace
+name: foo
+```
+
+and:
+
+```yaml
+kind: '*'
+namespace: foo
+```
+
+#### Web Role Editor
+
+To avoid confusion, in the web role editor, we would remove the ability to set
+and view `kubernetes_group` and `kubernetes_users` in favor of only allowing it
+to be controlled in yaml.
+When set in an existing role, similar to the existing `access` role, the UI can
+be disabled with a message saying:
+
+```text
+This role is too complex to be edited in the standard editor. To continue
+editing, go back to YAML editor.
+```
+
+The `kubernetes_resources` section currently "hidden" behind a button should be
+open by default.
+
+Alternatively, we could keep `kubernetes_group` under an "advanced" button,
+which would show what we see today by default as well as a link to a specific
+documentation page.
+
+#### CLI
+
+Using `tctl`, the fields will only be visible if they are set. Existing roles
+will have the existing values, even if already set to `kubernetes_master`,
+however, new roles can be created without the field and will be surfaced back
+without the field either.
+
+#### Direct enrollment
+
+When enrolling a cluster "directly" (i.e. by adding the `kubernetes_service`
+section in the `teleport.yaml` config), the initial setup of roles/rolebindings
+still needs to be performed calling the following script:
+[get-kubeconfig.sh](../examples/k8s-auth/get-kubeconfig.sh).
+
+It would be interesting to add a field to the config to allow specifying the
+default group that will be used for impersonation for advanced usecases to
+allow admins who prefer to rely on _Kubernetes_ RBAC instead of letting the
+agent be authoritative.
+
+#### Helm enrollment
+
+When using the Helm chart, nothing is expected to change, the chart sets up the
+service account for the agent, as well as the role/binding for impersonation.
+
+As _Helm_ is the favored way to get up and running, we may consider creating a
+dedicated group `teleport:master` or similar instead of using `system:master`.
+This would be cleaner but would rely on the Helm chart which would fracture the
+customer base and make support more difficult.
+
+It would be interesting to add an argument to the chart to allow specifying
+the default group that will be used for impersonation for advanced usecases.
```

Comment:
Note: changing the `system:masters` cluster role binding is something [we already support in the chart](https://goteleport.com/docs/reference/helm-reference/teleport-kube-agent/#adminclusterrolebinding) because of GKE autopilot clusters (they have some admission webhook rejecting direct usage of `system:masters`).

I would strongly prefer always creating and using our own group for consistency, especially to keep GKE users in the happy path. If the chart could tell the agent "here's your default group", we could safely make the chart always use `teleport:master` group while maintaining backward compatibility with the previous installations by defaulting to `system:masters` when the flag/config field is not set.